src/Ouroboros/Storage/Immutable/DB.hs

```diff
@@ -282,45 +296,192 @@ modifyInternalState db newEpoch action = do
                 action ns
             False -> action oldState
 
-getBinaryBlob :: forall m. (HasFSE m, MonadSTM m)
+
+-- | Loads an index file in memory.
+loadIndex :: forall m. HasFSE m
+          => ImmutableDB m
+          -> Epoch
+          -> ExceptT FsError m (U.Vector (Word64, Word64))
+loadIndex ImmutableDB{..} epoch = do
+    let indexFile = _dbFolder <> renderFile "index" epoch
+    fullIndex <- withFile indexFile ReadMode $ \hnd ->
+                     BL.toStrict . BS.toLazyByteString <$> readAll hnd mempty
+    let fullLen    = fromIntegral (BS.length fullIndex)
+        -- We need to divide by 2 at the end as we store tuples.
+        -- Example: We have stored only 1 slot in our index, so we have stored
+        -- an offset and an size for a total of 2 * sizeof(Word64) = 16 bytes.
+        -- To get the right size back we need to divide 16 by sizeof(Word64)
+        -- and finally by 2.
+        vectorSize = ceiling $
+            (fullLen / (fromIntegral indexEntrySizeBytes) :: Double) / 2.0
+        mkEntry ix =
+          let start = decodeIndexEntryAt (ix * indexEntrySizeBytes) fullIndex
+              end   = decodeIndexEntryAt ((ix + 1) * indexEntrySizeBytes) fullIndex
+              in (start, end - start)
+    return $ U.generate vectorSize mkEntry
+  where
+    -- Reads the data from the index file 64k bytes at the time.
+    readAll :: FsHandleE m
+            -> BS.Builder
+            -> ExceptT FsError m BS.Builder
+    readAll hnd acc = do
+        bytesRead <- hGet hnd 64000
+        let acc' = acc <> BS.byteString bytesRead
+        case BS.length bytesRead < 64000 of
+             True  -> return acc'
+             False -> readAll hnd acc'
+
+-- | An Iterator is a read handle which can be used to efficiently stream
+-- binary blobs across multiple epochs.
+data Iterator m = Iterator
+                { _it_current      :: TVar m (Epoch, RelativeSlot)
+                , _it_end          :: (Epoch, RelativeSlot)
+                , _it_epoch_handle :: TVar m (FsHandleE m)
+                , _it_epoch_index  :: TVar m (U.Vector (Word64, Word64))
+                -- ^ We load in memory the index file for the epoch we are
+                -- currently iterating on, as it's going to be small anyway
+                -- (usually ~150kb).
+                }
```

Comment:
Let's state (and test, possibly?) some invariants:

* `_it_next` never points to a non-existent block
* It never points to a block at or past the "next block" in the immutable DB.